openedx/core/lib/django_courseware_routers.py

```diff
@@ -32,16 +44,21 @@ def db_for_write(self, model, **hints):  # pylint: disable=unused-argument
         """
         Use the StudentModuleHistoryExtendedRouter.DATABASE_NAME if the model is StudentModuleHistoryExtended.
         """
-        if self._is_csmh(model):
+        if self._is_csm_h(model):
             return self.DATABASE_NAME
         else:
             return None
 
     def allow_relation(self, obj1, obj2, **hints):  # pylint: disable=unused-argument
         """
-        Disable relations if the model is StudentModuleHistoryExtended.
+        Manage relations if the model is StudentModuleHistoryExtended.
         """
-        if self._is_csmh(obj1) or self._is_csmh(obj2):
+        # Allow relation between CSM and CSMH (this cross-database relationship is declared with db_constraint=False).
+        if self._is_csm(obj1) and self._is_csm_h(obj2):
```

Comment:
I'm not super familiar with this functionality, but are we confident that we'll always get `CSM` as `obj1` and `CSMH` as `obj2`?

Reply:
This is my understanding based on the description at Django's [related_descriptors.py](https://github.com/django/django/blob/stable/2.2.x/django/db/models/fields/related_descriptors.py).

- The parent-to-child relationship it refers to corresponds to the `StudentModule`-to-`StudentModuleHistoryExtended` relationship via [this ForeignKey declaration](https://github.com/edx/edx-platform/blob/b489a4ecb12252099b1d249e588158ff038187a1/lms/djangoapps/coursewarehistoryextended/models.py#L43). 
- Given this relationship, the relation manager that validates the relationship for CSMH is the `ForwardManyToOneDescriptor`.
- The [__set__ docstring](https://github.com/django/django/blob/3ab5235d1dc94f7c8fe37a98c4e2c2337a5e5548/django/db/models/fields/related_descriptors.py#L194-L203 ) describes the arguments passed to `allow_relation` in this parent-to-child relationship.

Reply:
Merging as is now.  Will do a quick follow with a comment here to explain within the code.

Reply:
Added a comment in this follow-up PR: https://github.com/edx/edx-platform/pull/23568